common/types/hashable.go

```diff
@@ -0,0 +1,11 @@
+package types
+
+// A chain-agnostic generic interface to represent the following native types on various chains:
+// PublicKey, Address, Account, BlockHash, TxHash
+type Hashable interface {
+	MarshalText() (text []byte, err error)
+	UnmarshalText(text []byte) error
+	String() string
+	Equals(h Hashable) bool
```

Comment:
Will we only ever need direct equality? Or would a more general compare method be useful? (e.g. sorting)
```suggestion
	Compare(h Hashable) int
```
Also, we should document the expectation, if any, about whether h must be the same type or just any `Hashable` type. e.g. is it safe for implementations to cast? Can we ensure that is always safe to do?

Reply:
See the new version. Enforcing the type check via generics

Reply:
Compare for sorting etc, we can add when we need it. Not needed for now.